Introduce the "fix it" button to send error to the ai assistant

FadhlanR opened the pull request:
In this PR, I implemented the "Fix it" button. When clicked, it sends the error message and the related file to the AI assistant.

Syntax Error:


https://github.com/user-attachments/assets/32bdff6e-bdcc-4f4b-b4b1-2953d953499a


Error in playground panel:

https://github.com/user-attachments/assets/55a8de5d-7be5-4cfc-87c9-e1e9e9d71bdd

FadhlanR committed:
Create ai assitant panel service

FadhlanR committed:
Send error to ai assistant

FadhlanR committed:
Add file to fix

FadhlanR committed:
Add test

FadhlanR committed:
Fix test

FadhlanR committed:
Update design

jurgenwerk commented on packages/host/app/components/operator-mode/card-error-detail.gts:
```diff
@@ -54,22 +67,31 @@ export default class CardErrorDetail extends Component<Signature> {
-          </span>
+          <div class='title'>
+            <ExclamationCircle class='error-icon' />
+            {{if @headerText @headerText 'An error was encountered: '}}
```
```suggestion
            {{unless @headerText  'An error was encountered: '}}
```

jurgenwerk reviewed (commented):
1. I think "FixItButton" isn't the best name, it could be named in a way where it's clearer from the name what it does, for example `<SendToAiAssistant @fileToAttach={{fileToAttach}} @message={{message}},...>`

2. Consult with the design to get a design for the button (red isn't the best as it expresses negativity)

jurgenwerk commented on packages/host/app/components/operator-mode/fix-it-button.gts:
```diff
@@ -0,0 +1,101 @@
+
+    await this.matrixService.sendMessage(
+      this.matrixService.currentRoomId,
+      `In the attachment file, I encountered an error that needs fixing:\n\n${this.errorMessage}.`,
```
```suggestion
      `In the attached file, I encountered an error that needs fixing:\n\n${this.errorMessage}.`,
```

FadhlanR committed:
update checking

Co-authored-by: Matic Jurglič <matic@jurglic.si>

FadhlanR committed:
update test description

Co-authored-by: Matic Jurglič <matic@jurglic.si>

FadhlanR committed:
Merge branch 'main' into cs-7857-cs-8488-fix-it-button

FadhlanR committed:
lint fix

FadhlanR committed:
rename component

FadhlanR committed:
Update error display

FadhlanR committed:
Make new design works

FadhlanR commented:
I've implemented the new design. Here is the latest screenshot: 


https://github.com/user-attachments/assets/653bad97-3450-4c01-aff3-27974b16cc82



https://github.com/user-attachments/assets/c3106325-273d-4eef-8b00-a1e360e125f1

jurgenwerk commented:
In your video when you fix the file it says "playground is currently not supported for this file type". Is this a new error?

FadhlanR commented:
> In your video when you fix the file it says "playground is currently not supported for this file type". Is this a new error?

Not it's not, it's expected since the cursor position in monaco editor is not in the author card definition.

burieberry reviewed (commented):
Looking at the `boom-pet.gts` error card and its instance on local, something looks broken:
In interact mode, no details show up when expanding the box. Also, is the "Fix It" button supposed to appear here too?
<img width="500" alt="interact" src="https://github.com/user-attachments/assets/69e132a8-1cbe-42e6-bf48-e9ae0bdb766d" />

Code mode playground:
- Not seeing error details when expanding the box
- Hovering on the "Fix It" button causes the button to shift up. I see some css `transform: translateY(-1px);` why is this needed?
<img width="500" alt="error-details" src="https://github.com/user-attachments/assets/8665a866-c673-4995-9afc-cec2650a63ca" />

If the details is only when there's a stack trace, maybe we shouldn't display it if there's no stack trace.

jurgenwerk commented:
> Not it's not, it's expected since the cursor position in monaco editor is not in the author card definition.

It looks to me it's disruptive to the user's flow. I would expect the instance in the playground to show after error is fixed. If this is out of scope for this PR we should add a ticket I think

FadhlanR committed:
Fix tests

FadhlanR committed:
Fix host test

FadhlanR committed:
Another host test

FadhlanR committed:
Merge branch 'main' into cs-7857-cs-8488-fix-it-button

FadhlanR committed:
Merge resolve

FadhlanR committed:
Handle no stack trace

FadhlanR commented:
> Looking at the `boom-pet.gts` error card and its instance on local, something looks broken: In interact mode, no details show up when expanding the box. Also, is the "Fix It" button supposed to appear here too? <img alt="interact" width="500" src="https://private-user-images.githubusercontent.com/16160806/445183793-69e132a8-1cbe-42e6-bf48-e9ae0bdb766d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDc3MjI0NDEsIm5iZiI6MTc0NzcyMjE0MSwicGF0aCI6Ii8xNjE2MDgwNi80NDUxODM3OTMtNjllMTMyYTgtMWNiZS00MmU2LWJmNDgtZTlhZTBiZGI3NjZkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTIwVDA2MjIyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRhYmViNGUyN2RjOWRlY2M4MWNiZTZlMDRiNzQ5ODZiN2VlNjJhZjhlZGJhYmJiM2Q1NDUxODM0MWZiZGI2MWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.k6ZEjuMJzD5fUmuR7X_vNTmxErK9QmC-jMbKr4MF648">
> 
> Code mode playground:
> 
> * Not seeing error details when expanding the box
> * Hovering on the "Fix It" button causes the button to shift up. I see some css `transform: translateY(-1px);` why is this needed?
> 
> <img alt="error-details" width="500" src="https://private-user-images.githubusercontent.com/16160806/445186795-8665a866-c673-4995-9afc-cec2650a63ca.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDc3MjI0NDEsIm5iZiI6MTc0NzcyMjE0MSwicGF0aCI6Ii8xNjE2MDgwNi80NDUxODY3OTUtODY2NWE4NjYtYzY3My00OTk1LTlhZmMtY2VjMjY1MGE2M2NhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTIwVDA2MjIyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI0MWU3ZmRjMjhlZTNjN2Y0OTRiNmI3YmE1ZWU1MjM1ZjQ2YWViNmU1MjEwMGNmMGEwMTE1ZGRhODc5MmUwYzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.zhqOS7dn_xb53dAodJ7FfGtXErcRZjcwshw6KWftGtQ">
> If the details is only when there's a stack trace, maybe we shouldn't display it if there's no stack trace.

@burieberry For now this button will displayed for syntax error and the card error in playground panel. I removed the `transform: translateY(-1px);` and I fixed the blank details by adding better handling for cases where the stack trace is not provided. Instead of hiding the details area, I now display a message saying the stack trace is not available. Hiding the details area would also require hiding the "Show/Hide Details" button, and since there's only a copy button in the bottom-right, it looked weird.

With no stack trace:

<img width="638" alt="Screenshot 2025-05-20 at 13 43 39" src="https://github.com/user-attachments/assets/3a7f6629-a932-4149-933c-737172b99243" />

With stack trace:

https://github.com/user-attachments/assets/61985440-ba59-4bbd-bf36-e23510225965

FadhlanR commented:
> > Not it's not, it's expected since the cursor position in monaco editor is not in the author card definition.
> 
> It looks to me it's disruptive to the user's flow. I would expect the instance in the playground to show after error is fixed. If this is out of scope for this PR we should add a ticket I think

Maybe the screenshot wasn’t very clear—here’s the full screenshot before and after the error. As you can see in the video below, the playground panel is not displayed because the editing is happening in the JavaScript class, not in the card or field definition. That’s why the playground panel doesn't appear.


https://github.com/user-attachments/assets/c3061647-36ae-4683-abd7-413d2898ca36

FadhlanR committed:
box shadow

FadhlanR committed:
Move view in code button

jurgenwerk commented:
I am trying to test this and the button is not showing for me: 
<img width="1333" alt="image" src="https://github.com/user-attachments/assets/0ab14df7-f78d-4ab0-9e38-0628c88329c3" />

jurgenwerk commented on packages/host/app/components/operator-mode/error-display.gts:
```diff
@@ -0,0 +1,250 @@
+        {{#if
+          (and
+            (bool @fileToAttach)
+            (not this.operatorModeStateService.aiAssistantOpen)
```
Why not show the fix button if the ai panel is open?

FadhlanR replied:
To make more room: https://discord.com/channels/584043165066199050/584061276544827432/1372990016581799976.

FadhlanR replied:
Or do you think it is better to not hide it?

jurgenwerk replied:
I'm confused, room for which thing? I think the link you gave is talking about "Ask AI" input not the "Fix it" button

Yes, I don't think hiding this button when AI panel is open makes any sense. We need that button so that you can use it to quickly fix an error when you are working with the assistant.

FadhlanR replied:
Ah I thought "Ask AI" is the "Fix it" button. I'll update it.

FadhlanR replied:
@jurgenwerk updated, please check it again

burieberry commented on packages/host/app/components/operator-mode/error-display.gts:
```diff
@@ -0,0 +1,250 @@
+
+    <style scoped>
+      .error-display {
+        background: #ffba00;
```
you can add this color to `variables.css` as `boxel-warning-200`

burieberry commented on packages/boxel-ui/addon/src/icons.gts:
```diff
@@ -20,6 +20,7 @@ import DropdownArrowFilled from './icons/dropdown-arrow-filled.gts';
 import DropdownArrowUp from './icons/dropdown-arrow-up.gts';
 import Exclamation from './icons/exclamation.gts';
 import ExclamationCircle from './icons/exclamation-circle.gts';
+import ExclamationTriangleFill from './icons/exclamation-triangle-fill.gts';
```
There's a very similar icon `warning.svg` already existing here. Does that icon work?

burieberry commented on packages/host/app/components/operator-mode/card-error-detail.gts:
```diff
@@ -1,146 +1,57 @@
-      }
-      .error-detail :deep(.accordion-item) {
-        height: auto;
+        margin: var(--boxel-sp);
```
the `margin` in line 48 is overriding the `margin-top` in line 46. as a result, this component does not always appear at the bottom of the page.

burieberry commented on packages/host/app/components/operator-mode/error-display.gts:
```diff
@@ -0,0 +1,250 @@
+      <div class='error-actions'>
+        <CopyButton @textToCopy={{@message}} @width='16px' @heigth='16px' />
+        <button
+          class='toggle-details-button'
```
Can probably remove a lot of extra css by using `<Button @kind='text-only' @size='extra-small' ...`

burieberry commented on packages/host/app/components/operator-mode/send-error-to-ai-assistant.gts:
```diff
@@ -0,0 +1,103 @@
+    <Button
+      class='send-error-to-ai-assistant'
+      @kind='primary'
+      @size='small'
```
maybe this is better: `@kind='secondary-dark' @size='extra-small'`

burieberry commented on packages/host/app/components/operator-mode/send-error-to-ai-assistant.gts:
```diff
@@ -0,0 +1,103 @@
+        display: inline-flex;
+        align-items: center;
+        gap: 6px;
+        transition: all 0.2s ease;
```
can use more specific transition properties if this is needed

burieberry reviewed (commented):
Made some additional suggestions, but overall it looks good to me

FadhlanR committed:
Address feedback

FadhlanR committed:
Merge branch 'main' into cs-7857-cs-8488-fix-it-button

burieberry reviewed (approved):
I think I saw a `console.log` leftover in the code, otherwise looks good

FadhlanR committed:
remove unnecessary console log

FadhlanR committed:
Merge branch 'main' into cs-7857-cs-8488-fix-it-button
